lib/application.js

```diff
@@ -404,30 +405,133 @@ app.enableAuth = function(options) {
+  const User = self.registry.findModel('User');
+  const models = self.models();
 
-    if (Model === User || Model.prototype instanceof User) {
-      scheduleVerification(Model, verifyUserRelations);
-    }
+  verifyUserModelsSetup().then(()=> {
+    verifyRelationsSetup();
   });
 
+  function verifyUserModelsSetup() {
+    let userModels = models.filter(model => {
+      return (model === User || model.prototype instanceof User) ||
+        (model === AccessToken || model.prototype instanceof AccessToken);
+    });
+
+    // we use Promise.reflect() here to let all the tests go without throwing,
+    // this way we can get all the logs for all models we are checking
+    return Promise.map(userModels, (model) => {
+      return scheduleVerification(model, hasMultipleUserModelsConfig).reflect();
+    })
+    .then((inspections) => {
+      // proceed to next checkups if no error
+      let hasErrors = inspections.filter(inspection => {
+        return inspection.isRejected();
+      });
+      if (!hasErrors.length) {
+        return;
+      }
+
+      // else log and eventually kill the app
+      console.warn('Application setup is inconsistent: some models are set to use ' +
+        'the Multiple user models configuration while some other models are set to ' +
+        'use the Single user model configuration. The model config is listed below.');
+
+      // detailed logs
+      inspections.forEach(inspection => {
+        let modelSetup = (inspection.value() || inspection.reason()).modelSetup;
+        console.warn(
+          'Model %j of type %j is set to use the %j user model config',
+          modelSetup.model,
+          modelSetup.instanceOf,
+          modelSetup.hasMultipleUserModelsConfig ? 'Multiple' : 'Single'
+        );
+      });
+
+      delete self.hasMultipleUserModelsConfig;
+
+      if (!dismissRejection) {
+        throw new Error('Application setup is inconsistent, please see the log for more ' +
+          'information');
+      }
+    });
+  }
+
+  function verifyRelationsSetup() {
+    models.forEach(Model => {
+      if (Model === AccessToken || Model.prototype instanceof AccessToken) {
+        scheduleVerification(Model, verifyAccessTokenRelations);
+      }
+
+      if (Model === User || Model.prototype instanceof User) {
+        scheduleVerification(Model, verifyUserRelations);
+      }
+    });
+  }
+
+  // instant or scheduled Model verifications, as a promise
   function scheduleVerification(Model, verifyFn) {
-    if (Model.dataSource) {
-      verifyFn(Model);
-    } else {
-      Model.on('attached', () => verifyFn(Model));
+    return new Promise((resolve, reject) => {
+      if (Model.dataSource) {
+        try {
+          resolve(verifyFn(Model));
+        } catch (e) {
+          reject(e);
+        }
+      } else {
+        Model.on('attached', () => {
+          scheduleVerification(Model, verifyFn);
+        });
+      }
+    });
+  }
```

Comment:
i needed to change @bajtos first implem to add async support as we now expect on this function to resolve when done, so we can use it to first check user models config and then check how the relations are set up